### PR TITLE
TIP-898: Setup default properties

### PR DESCRIPTION
**Pull request by anaelChardan**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/UserManagement/Component/Factory/UserFactory.php [hunk 1]
```diff
@@ -81,7 +87,9 @@ public function create()
             $user->addGroup($group);
         }
 
-        return $user;
+        return array_reduce($this->defaultProperties, function ($user, DefaultProperty $defaultProperty) {
```
Comment (arnolanglade) [inline src/Akeneo/UserManagement/Component/Factory/UserFactory.php]: yeah... I don't think this a useful use case for array_reduce.

File: tests/back/UserManagement/Specification/Component/Factory/SimpleDefaultPropertySpec.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    function it_is_a_default_property()
+    {
+        $this->shouldImplement(DefaultProperty::class);
+    }
+
+    function it_mutates_the_user()
+    {
+        $user = new User();
+
+        $this->mutate($user)->shouldReturnAnInstanceOf(UserInterface::class);
```
Comment (arnolanglade) [inline tests/back/UserManagement/Specification/Component/Factory/SimpleDefaultPropertySpec.php]: nope `shouldReturn(new User()->addProperty('a_default_value', 'a_property'))`

File: src/Akeneo/UserManagement/Component/Factory/DefaultProperty.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+use Akeneo\UserManagement\Component\Model\UserInterface;
+
+/**
+ * Interface to mutate User's default property
+ *
+ * @author    Anael Chardan <anael.chardan@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface DefaultProperty
```
Comment (arnolanglade) [inline src/Akeneo/UserManagement/Component/Factory/DefaultProperty.php]: this interface is useless

File: src/Akeneo/UserManagement/Component/Factory/SimpleDefaultProperty.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+use Akeneo\UserManagement\Component\Model\UserInterface;
+
+/**
+ * Mutates simple default property of a user
+ *
+ * @author    Anael Chardan <anael.chardan@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class SimpleDefaultProperty implements DefaultProperty
```
Comment (arnolanglade) [inline src/Akeneo/UserManagement/Component/Factory/SimpleDefaultProperty.php]: you should rename it to DefaultProperty

Comment (arnolanglade) [reply src/Akeneo/UserManagement/Component/Factory/SimpleDefaultProperty.php]: ok I understand why!

File: src/Akeneo/UserManagement/Component/Factory/DefaultProperty.php (hunk 3 above)
Comment (arnolanglade) [reply src/Akeneo/UserManagement/Component/Factory/DefaultProperty.php]: ok I understand why!

File: src/Akeneo/UserManagement/Component/Factory/UserFactory.php (hunk 1 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Component/Factory/UserFactory.php]: It's the perfect case for an array reduce ;) An array_reduce is an foldable thing that means you reduce something to something else it is not only for doing sum or product in terms of integer but it is for everything, here is kinda a cartesian product ;)

File: src/Akeneo/UserManagement/Component/Factory/DefaultProperty.php (hunk 3 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Component/Factory/DefaultProperty.php]: So what type do you accept in the Factory?

File: src/Akeneo/UserManagement/Component/Factory/SimpleDefaultProperty.php (hunk 4 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Component/Factory/SimpleDefaultProperty.php]: We'll see :)

File: tests/back/UserManagement/Specification/Component/Factory/SimpleDefaultPropertySpec.php (hunk 2 above)
Comment (anaelChardan) [reply tests/back/UserManagement/Specification/Component/Factory/SimpleDefaultPropertySpec.php]: Problem with the salt which is generated at the instantiation :/

**Pull request merged**